Refactoring/gcp/iam

Remi05 opened the pull request:
Migrated GCP's IAM service to the new architecture.  

Note that `gcp/resources/projects.py` and `gcp/facade/base.py` are the same as in the [Stackdriver logging PR](https://github.com/nccgroup/ScoutSuite/pull/229); they are simply shared by all services and will appear in all GCP refactoring PRs until one of the services is merged.

Remi05 commented:
The service seems to work fine but the resource counts have changed; In the previous implementation, IAM showed a count of 8 resources where as the new implementation shows 4 resources. In both cases, there are 4 service accounts, which is shown correctly in the top navigation bar for both implementations. I thought it might be due to the keys and bindings counts but they do not seem to be set, even in the previous implementation.

Before:
![iam_beforerefactoring](https://user-images.githubusercontent.com/12770956/53931907-29e10c80-4065-11e9-8645-0f1ed8fe73c9.png)

After:
![iam_afterrefactoring](https://user-images.githubusercontent.com/12770956/53931912-31a0b100-4065-11e9-9961-1939f4ecf838.png)

Remi05 commented:
The build doesn't pass for Python 3.5 because of formatted strings using the `f'Text {my_variable}'` syntax which was introduced in Python 3.6. Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?

j4v commented:
> Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?

Well asyncio is available from 3.5+, but 3.6+ should be the standard 3.x on most distros. If there's a benefit to only supporting 3.6+ that's something we could consider. I'm seeing a lot of changes happening in python 3.x so if we're along for the ride (and everyone will have to be from 2020 onward) then we might as well just accept that having outdated versions of python is not acceptable anymore.

Aboisier commented:
> > Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?
> 
> Well asyncio is available from 3.5+, but 3.6+ should be the standard 3.x on most distros. If there's a benefit to only supporting 3.6+ that's something we could consider. I'm seeing a lot of changes happening in python 3.x so if we're along for the ride (and everyone will have to be from 2020 onward) then we might as well just accept that having outdated versions of python is not acceptable anymore.

I'm not sure formatted string literals is a good enough reason to deprecate 3.5... but at the same time, it seems like there's barely anyone using 3.5. 

![image](https://user-images.githubusercontent.com/6698529/53960649-d1396000-40b4-11e9-9ce7-86e0da4588b3.png)

Aboisier commented on ScoutSuite/providers/gcp/configs/services.py:
```diff
@@ -20,11 +22,14 @@ def __init__(self, metadata=None, thread_config=4, projects=None, **kwargs):
         self.cloudsql = CloudSQLConfig(thread_config=thread_config)
         self.computeengine = ComputeEngineConfig(thread_config=thread_config)
-        self.iam = IAMConfig(thread_config=thread_config)
+        self.iam = IAM(gcp_facade, iam_facade)
```
Same comment as with the Stack Driver Logging PR, I'd merge the two facades.

Aboisier commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,29 @@
+
+class IAMFacade:
+    def __init__(self):
+        self._iam_client = discovery.build('iam', 'v1', cache_discovery=False, cache=MemoryCache())
```
[Same comment](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263377080), I don't think this is the right place to do HTTP requests.

Aboisier commented on ScoutSuite/providers/gcp/resources/iam/service_accounts.py:
```diff
@@ -0,0 +1,38 @@
+    async def fetch_all(self):
+        raw_service_accounts = await self.iam_facade.get_service_accounts(self.project_id)
+        for raw_service_account in raw_service_accounts:
+            sa_id, service_account = self._parse_service_account(raw_service_account)
```
I'd rename `sa_id` for `service_account_id`.

Aboisier commented on ScoutSuite/providers/gcp/resources/iam/service_accounts.py:
```diff
@@ -0,0 +1,38 @@
+        await self._fetch_children()
+    
+    async def _fetch_children(self):
+        for sa_id, service_account in self.items():
```
Same here, I'd rename `sa_id` to `service_account_id`.

Aboisier commented on ScoutSuite/providers/gcp/resources/iam/service_accounts.py:
```diff
@@ -0,0 +1,38 @@
+                await child.fetch_all()
+                self[sa_id][child_name] = child
+
+    def _parse_service_account(self, raw_service_account):
```
[Same as here](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263382611), would rename to `_parse`.

Aboisier commented on ScoutSuite/providers/gcp/resources/iam/bindings.py:
```diff
@@ -0,0 +1,18 @@
+            binding_id, binding = self._parse_binding(raw_binding)
+            self[binding_id] = binding
+
+    def _parse_binding(self, raw_binding):
```
[Same](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263382611), would rename to `_parse`.

Aboisier commented on ScoutSuite/providers/gcp/resources/iam/keys.py:
```diff
@@ -0,0 +1,23 @@
+            key_id, key = self._parse_key(raw_key)
+            self[key_id] = key
+
+    def _parse_key(self, raw_key):
```
[Same](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263382611), would rename to `_parse`.

Aboisier commented on ScoutSuite/providers/gcp/facade/facade.py:
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
```
[Same as here](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263374292), would remove this line. I won't point out every occurrence but applies to all the other files.

Aboisier commented on ScoutSuite/providers/gcp/facade/facade.py:
```diff
@@ -0,0 +1,18 @@
+        self._resourcemanager_client = discovery.build('cloudresourcemanager', 'v1', cache_discovery=False, cache=MemoryCache())
+
+    # TODO: Make truly async    
+    async def get_projects(self):
```
I think you could generalize this method. There is code duplication between this and [this](https://github.com/nccgroup/ScoutSuite/pull/230/files#diff-ffea74d1a65db135e82d200249fd7e5eR21).

Remi05 replied:
Yeah, I had noticed but I wasn't sure how to do it cleanly, I think I found a decent way. Let me know what you think!

Aboisier replied:
Yup, that's exactly what I had in mind. You'll have to make this async though (like [this commit](https://github.com/nccgroup/ScoutSuite/pull/224/commits/25967e57b29d9c66af0daf339162aa5e0492e3ea)). Michael spent a few days looking for the performance bottleneck, and I spent a day, and it was that; it makes a huge difference.

Remi05 replied:
I'm aware, that's what the comment was for, I was waiting for Michael's PR to be merged before adding it! 😉

j4v commented:
> it seems like there's barely anyone using 3.5

:laughing: 

There's 2 types of people, those who install dependencies on their host system and burn in hell, and those who use virtual environments. The later shouldn't care what version of 3.x is used...

Remi05 commented:
> > > Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?
> > 
> > 
> > Well asyncio is available from 3.5+, but 3.6+ should be the standard 3.x on most distros. If there's a benefit to only supporting 3.6+ that's something we could consider. I'm seeing a lot of changes happening in python 3.x so if we're along for the ride (and everyone will have to be from 2020 onward) then we might as well just accept that having outdated versions of python is not acceptable anymore.
> 
> I'm not sure formatted string literals is a good enough reason to deprecate 3.5... but at the same time, it seems like there's barely anyone using 3.5.
> 
> ![image](https://user-images.githubusercontent.com/6698529/53960649-d1396000-40b4-11e9-9ce7-86e0da4588b3.png)

I wouldn't deprecated Python 3.5 over formatted strings, especially since you can format them differently, but if we were already going to, I would have kept this cleaner syntax 😉

Remi05 commented:
I replaced formatted string using the `f'Text {variable}'` with `'{} text {}'.format(var1, var2)` to remain compatible with Python 3.5!

misg commented on ScoutSuite/providers/gcp/facade/facade.py:
```diff
@@ -0,0 +1,12 @@
+
+class Facade:
```
Is this an attempt to make an abstract interface for all cloud provider facades? Seeing how really different they are, I'm not sure we can really make them homogeneous...

Remi05 replied:
No it was simply for GCP, it avoids duplicating the `_get_client()` method, I can give it another name though

Aboisier replied:
For AWS, I called my base facade  `AWSBaseFacade`. I think it's less confusing than `Facade`.

Remi05 replied:
Renamed it to `GCPBaseFacade`!

misg commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,30 @@
+from ScoutSuite.providers.gcp.utils import MemoryCache
+from ScoutSuite.providers.gcp.facade.facade import Facade
+from ScoutSuite.providers.gcp.facade.utils import GCPFacadeUtils
+
```
```suggestion
from ScoutSuite.providers.utils import run_concurrently

```

misg commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,30 @@
+    async def get_bindings(self, project_id, service_account_email):
+        resource = 'projects/{}/serviceAccounts/{}'.format(project_id, service_account_email)
+        iam_client = self._get_client()
+        response = iam_client.projects().serviceAccounts().getIamPolicy(resource=resource).execute()
```
```suggestion
        response = await run_concurrently(
                lambda: iam_client.projects().serviceAccounts().getIamPolicy(resource=resource).execute()
        )
```

misg commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,30 @@
+    async def get_keys(self, project_id, service_account_email):
+        name = 'projects/{}/serviceAccounts/{}'.format(project_id, service_account_email)
+        iam_client = self._get_client()
+        response = iam_client.projects().serviceAccounts().keys().list(name=name).execute()
```
```suggestion
        response = await run_concurrently(
                lambda: iam_client.projects().serviceAccounts().keys().list(name=name).execute()
        )
```

misg commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,30 @@
+        response = iam_client.projects().serviceAccounts().keys().list(name=name).execute()
+        return response.get('keys', [])
+
+    # TODO: Make truly async
```
Already truly async afaik, unless there's an API call behind the assignment of `request` and/or `service_accounts_group`?

Remi05 replied:
That was before I extracted the `get_all()` loop, I'll update the comment 😉

misg commented on ScoutSuite/providers/gcp/resources/iam/service_accounts.py:
```diff
@@ -0,0 +1,38 @@
+        await self._fetch_children()
+    
+    async def _fetch_children(self):
+        for service_account_id, service_account in self.items():
```
You'll need to parallelize this loop (see https://github.com/nccgroup/ScoutSuite/pull/229/files#r263374926) and probably make a refactoring of `fetch_children` as well (would be probably worth it to define a `GCPCompositeResources` to do so, like `AWSCompositeResources` and `AzureCompositeResources` ([here](https://github.com/nccgroup/ScoutSuite/blob/28e6d1f3e4bd4b068c0b0caf31eb718743919d3c/ScoutSuite/providers/aws/resources/resources.py#L25) and [there](https://github.com/nccgroup/ScoutSuite/blob/28e6d1f3e4bd4b068c0b0caf31eb718743919d3c/ScoutSuite/providers/azure/resources/resources.py#L8))).

Remi05 replied:
Done!

Aboisier commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,30 @@
+    async def get_service_accounts(self, project_id):
+        name = 'projects/{}'.format(project_id)     
+        iam_client = self._get_client()
+        request = iam_client.projects().serviceAccounts().list(name=name)
```
I think this should be parallelized!

```suggestion
        request = await run_concurrently(
                iam_client.projects().serviceAccounts().list(name=name)
        )
```

Remi05 replied:
`request` and `services_accounts_group` simply build the request from my understanding. The actual fetching is done in `GCPFacadeUtils.get_all()` which is already parallelized .

Aboisier commented on ScoutSuite/providers/gcp/facade/iam.py:
```diff
@@ -0,0 +1,30 @@
+        name = 'projects/{}'.format(project_id)     
+        iam_client = self._get_client()
+        request = iam_client.projects().serviceAccounts().list(name=name)
+        service_accounts_group = iam_client.projects().serviceAccounts()
```
Should that be parallelized too?
```suggestion
        service_accounts_group = await run_concurrently(
                iam_client.projects().serviceAccounts()
        )
```

Remi05 commented:
Fixed UI and added left menu for projects.

Before : 

![iam_before](https://user-images.githubusercontent.com/12770956/55370524-bea22300-54c8-11e9-920b-d986aa8bda4a.png)

After : 

![iam_after](https://user-images.githubusercontent.com/12770956/55370534-c5c93100-54c8-11e9-9700-3e263c62754b.png)

Aboisier commented on ScoutSuite/providers/gcp/resources/iam/keys.py:
```diff
@@ -0,0 +1,22 @@
+
+    def _parse_key(self, raw_key):
+        key_dict = {}
+        key_dict['id'] = raw_key['name'].split('/')[-1]
```
Can we add a small comment here explaining where this comes from?

Remi05 replied:
Done!
